### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by ninahakansson**
Refactor code and add tests

Refactor gac2pps_lib and mersi22pps_lib similar to how seviri2pps_lib was refactored in PR17.
Add tests of mersi22pps_lib and gac2pps_lib.

Change angle data set names from imageX to sunzenith, satzenith, etc.

Added a test for gac2pps using a cropped gac file.

### CODE REVIEW

File: level1c4pps/__init__.py [hunk 1]
```diff
@@ -55,6 +88,146 @@ def dt64_to_datetime(dt64):
...
+        encoding[name] = enc
+    return encoding
+
+
+def get_band_encoding(dataset, bandnames, pps_tagnames, chunks=None):
+    """Get netcdf encoding for a datasets."""
+    name = dataset.attrs['name']
+    id_tag = dataset.attrs.get('id_tag', None)
+    if id_tag is not None:
+        if 'ch_tb' in id_tag:
```
Comment (mraspaud) [inline level1c4pps/__init__.py]: would `startswith` be more exact ?

File: level1c4pps/__init__.py [hunk 2]
```diff
@@ -55,6 +88,146 @@ def dt64_to_datetime(dt64):
...
+    platform_name = scene.attrs['platform']
+    orbit_number = int(scene.attrs['orbit_number'])
+    filename = os.path.join(
+        out_path,
+        "S_NWC_{:s}_{:s}_{:05d}_{:s}Z_{:s}Z.nc".format(
+            instrument,
+            platform_name.lower().replace('-', ''),
+            orbit_number,
+            datetime.strftime(dt64_to_datetime(start_time), '%Y%m%dT%H%M%S%f')[:-5],
+            datetime.strftime(dt64_to_datetime(end_time), '%Y%m%dT%H%M%S%f')[:-5]))
```
Comment (mraspaud) [inline level1c4pps/__init__.py:230]: Are you aware of the excellent `trollsift` library which is made just for that ? 
https://trollsift.readthedocs.io/en/latest/usage.html#parser

File: level1c4pps/__init__.py [hunk 3]
```diff
@@ -55,6 +88,146 @@ def dt64_to_datetime(dt64):
...
+    try:
+        del scene['lat'].coords['acq_time']
+        del scene['lon'].coords['acq_time']
+    except KeyError:
+        pass
+
+
+def update_angle_attributes(scene, start_time):
+    """Set and delete angle attributes."""
+    for angle in ['sunzenith', 'satzenith', 'azimuthdiff', 'sunazimuth', 'satazimuth']:
```
Comment (mraspaud) [inline level1c4pps/__init__.py]: Isn't that `PPS_ANGLE_TAGS` ?

File: level1c4pps/__init__.py [hunk 4]
```diff
@@ -55,6 +88,146 @@ def dt64_to_datetime(dt64):
...
+        if angle not in scene.keys() and angle in ['sunazimuth', 'satazimuth']:
+            # azimuth angles not always there
+            continue
+        scene[angle].attrs['id_tag'] = angle
+        scene[angle].attrs['name'] = angle
+        scene[angle].attrs['coordinates'] = 'lon lat'
+        scene[angle].attrs['long_name'] = ANGLE_ATTRIBUTES['long_name'][angle]
+        scene[angle].attrs['valid_range'] = ANGLE_ATTRIBUTES['valid_range'][angle]
+        scene[angle].attrs['standard_name'] = ANGLE_ATTRIBUTES['standard_name'][angle]
```
Comment (mraspaud) [inline level1c4pps/__init__.py:186]: You could make a loop here

File: level1c4pps/seviri2pps_lib.py [hunk 5]
```diff
@@ -278,103 +281,37 @@ def add_ancillary_datasets(scene, lons, lats, sunz, satz, azidiff,
...
-    scene['azimuthdiff'].attrs['name'] = "image13"
 
     # Some common attributes
     for angle in ['azimuthdiff', 'satzenith', 'sunzenith']:
         scene[angle].attrs['units'] = 'degree'
+        scene[angle].attrs['id_tag'] = angle
+        scene[angle].attrs['name'] = angle
+        scene[angle].attrs['long_name'] = ANGLE_ATTRIBUTES['long_name'][angle]
+        scene[angle].attrs['valid_range'] = ANGLE_ATTRIBUTES['valid_range'][angle]
+        scene[angle].attrs['standard_name'] = ANGLE_ATTRIBUTES['standard_name'][angle]
```
Comment (mraspaud) [inline level1c4pps/seviri2pps_lib.py]: Isn't that the same as in init ?

**Review by mraspaud — CHANGES_REQUESTED** [review]
Lots of factorization, really nice! A few small things I noted here.

File: level1c4pps/__init__.py (hunk 2 above)
Comment (ninahakansson) [reply level1c4pps/__init__.py:230]: Now I am, thanks!

File: level1c4pps/__init__.py (hunk 1 above)
Comment (ninahakansson) [reply level1c4pps/__init__.py]: Yes!

File: level1c4pps/__init__.py (hunk 3 above)
Comment (ninahakansson) [reply level1c4pps/__init__.py]: Yes, thanks!

File: level1c4pps/seviri2pps_lib.py (hunk 5 above)
Comment (ninahakansson) [reply level1c4pps/seviri2pps_lib.py]: Almost, I will fix that.

**Comment by sfinkens** [comment]
Sorry, I was too slow here

**Pull request merged**